<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>R6 Class Representing a Reference Genome — ref_genome • jackalope</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="R6 Class Representing a Reference Genome — ref_genome" />
<meta property="og:description" content="Interactive wrapper for a pointer to a C++ object that stores reference genome
information." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lucasnell/jackalope">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>R6 Class Representing a Reference Genome</h1>
    <small class="dont-index">Source: <a href='https://github.com/lucasnell/jackalope/blob/master/R/aaa-classes.R'><code>R/aaa-classes.R</code></a></small>
    <div class="hidden name"><code>ref_genome.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interactive wrapper for a pointer to a C++ object that stores reference genome
information.</p>
    </div>



    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This class should NEVER be created using <code>ref_genome$new</code>.
Only use <code>read_fasta</code> or <code>create_genome</code>.
Because this class wraps a pointer to a C++ object, there are no fields to
manipulate directly.
All manipulations are done through this class's methods.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='read_fasta.html'>read_fasta</a></code> <code><a href='create_genome.html'>create_genome</a></code></p></div>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>ref_genome$new()</code></a></p></li>
<li><p><a href='#method-print'><code>ref_genome$print()</code></a></p></li>
<li><p><a href='#method-ptr'><code>ref_genome$ptr()</code></a></p></li>
<li><p><a href='#method-n_chroms'><code>ref_genome$n_chroms()</code></a></p></li>
<li><p><a href='#method-sizes'><code>ref_genome$sizes()</code></a></p></li>
<li><p><a href='#method-chrom_names'><code>ref_genome$chrom_names()</code></a></p></li>
<li><p><a href='#method-chrom'><code>ref_genome$chrom()</code></a></p></li>
<li><p><a href='#method-gc_prop'><code>ref_genome$gc_prop()</code></a></p></li>
<li><p><a href='#method-nt_prop'><code>ref_genome$nt_prop()</code></a></p></li>
<li><p><a href='#method-set_names'><code>ref_genome$set_names()</code></a></p></li>
<li><p><a href='#method-clean_names'><code>ref_genome$clean_names()</code></a></p></li>
<li><p><a href='#method-add_chroms'><code>ref_genome$add_chroms()</code></a></p></li>
<li><p><a href='#method-rm_chroms'><code>ref_genome$rm_chroms()</code></a></p></li>
<li><p><a href='#method-merge_chroms'><code>ref_genome$merge_chroms()</code></a></p></li>
<li><p><a href='#method-filter_chroms'><code>ref_genome$filter_chroms()</code></a></p></li>
<li><p><a href='#method-replace_Ns'><code>ref_genome$replace_Ns()</code></a></p></li>
</ul>
<p><hr>
<a id="method-new"></a></p><h3>Method <code>new()</code></h3>
<p>Do NOT use this; only use <code>read_fasta</code> or <code>create_genome</code> to make a
new <code>ref_genome</code>.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$new(genome_ptr)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>genome_ptr</code></dt><dd><p>An <code>externalptr</code> object pointing to a C++ object that stores
the information about the reference genome.</p></dd>

</dl><p></div></p>
<p><hr>
<a id="method-print"></a></p><h3>Method <code><a href='https://rdrr.io/r/base/print.html'>print()</a></code></h3>
<p>Print a <code>ref_genome</code> object.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$print()</pre><p></div></p>

<p><hr>
<a id="method-ptr"></a></p><h3>Method <code>ptr()</code></h3>
<p>View pointer to underlying C++ object (this is not useful to end users).</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$ptr()</pre><p></div></p>

<h3>Returns</h3>
<p>An <code>externalptr</code> object.</p>
<p><hr>
<a id="method-n_chroms"></a></p><h3>Method <code>n_chroms()</code></h3>
<p>View number of chromosomes.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$n_chroms()</pre><p></div></p>

<h3>Returns</h3>
<p>Integer number of chromosomes.</p>
<p><hr>
<a id="method-sizes"></a></p><h3>Method <code>sizes()</code></h3>
<p>View chromosome sizes.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$sizes()</pre><p></div></p>

<h3>Returns</h3>
<p>Integer vector of chromosome sizes.</p>
<p><hr>
<a id="method-chrom_names"></a></p><h3>Method <code>chrom_names()</code></h3>
<p>View chromosome names.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$chrom_names()</pre><p></div></p>

<h3>Returns</h3>
<p>Character vector of chromosome names.</p>
<p><hr>
<a id="method-chrom"></a></p><h3>Method <code>chrom()</code></h3>
<p>View one reference chromosome.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$chrom(chrom_ind)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A single string representing the chosen chromosome's DNA sequence.</p>
<p><hr>
<a id="method-gc_prop"></a></p><h3>Method <code>gc_prop()</code></h3>
<p>View GC proportion for part of one reference chromosome.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$gc_prop(chrom_ind, start, end)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>start</code></dt><dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>

<dt><code>end</code></dt><dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is either <code>G</code> or <code>C</code>.</p>
<p><hr>
<a id="method-nt_prop"></a></p><h3>Method <code>nt_prop()</code></h3>
<p>View nucleotide content for part of one reference chromosome</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$nt_prop(nt, chrom_ind, start, end)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>nt</code></dt><dd><p>Which nucleotide to calculate the proportion that the DNA
sequence is made of. Must be one of <code>T</code>, <code>C</code>, <code>A</code>, <code>G</code>, or <code>N</code>.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>start</code></dt><dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>

<dt><code>end</code></dt><dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is <code>nt</code>.</p>
<p><hr>
<a id="method-set_names"></a></p><h3>Method <code>set_names()</code></h3>
<p>Change chromosome names.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$set_names(new_names)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>new_names</code></dt><dd><p>Vector of new names to use. This must be the same length as
the number of current names.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 10)
ref$set_names(c("a", "b", "c", "d"))

</pre><p></div></p>

<p><hr>
<a id="method-clean_names"></a></p><h3>Method <code>clean_names()</code></h3>
<p>Clean chromosome names, converting <code>" :;=%,\\|/\"\'"</code> to <code>"_"</code>.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$clean_names()</pre><p></div></p>

<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 10)
ref$set_names(c("a:", "b|", "c;", "d'"))
ref$clean_names()

</pre><p></div></p>

<p><hr>
<a id="method-add_chroms"></a></p><h3>Method <code>add_chroms()</code></h3>
<p>Add one or more chromosomes.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$add_chroms(new_chroms, new_names = NULL)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>new_chroms</code></dt><dd><p>Character vector of DNA strings representing new chromosomes.</p></dd>

<dt><code>new_names</code></dt><dd><p>Optional character vector of names for the new chromosomes.
It should be the same length as <code>new_chroms</code>.
If <code>NULL</code>, new names will be automatically generated. Defaults to <code>NULL</code>.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 10)
ref$add_chroms("TCAGTCAG")
</pre><p></div></p>

<p><hr>
<a id="method-rm_chroms"></a></p><h3>Method <code>rm_chroms()</code></h3>
<p>Remove one or more chromosomes by name</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$rm_chroms(chrom_names)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>chrom_names</code></dt><dd><p>Vector of the name(s) of the chromosome(s) to remove.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 10)
ref$set_names(c("a", "b", "c", "d"))
ref$rm_chroms("b")

</pre><p></div></p>

<p><hr>
<a id="method-merge_chroms"></a></p><h3>Method <code>merge_chroms()</code></h3>
<p>Merge all chromosomes into one after first shuffling their order.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$merge_chroms()</pre><p></div></p>

<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 10)
ref$merge_chroms()

</pre><p></div></p>

<p><hr>
<a id="method-filter_chroms"></a></p><h3>Method <code>filter_chroms()</code></h3>
<p>Filter chromosomes by size or for a proportion of total bases.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$filter_chroms(threshold, method)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>threshold</code></dt><dd><p>Number used as a threshold. If <code>method == "size"</code>,
then this is the minimum length of a chromosome that will remain after
filtering.
If <code>method == "prop"</code>, chromosomes are first size-sorted, then
the largest <code>N</code> chromosomes are retained that allow at least
<code>threshold * sum(&lt;all chromosome sizes&gt;)</code> base pairs remaining after
filtering.</p></dd>

<dt><code>method</code></dt><dd><p>String indicating which filter method to use: chromosome size
(<code>method = "size"</code>) or proportion of total bases (<code>method = "prop"</code>).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<h3>Examples</h3>
<p><div class="r example copy"></p><pre>ref &lt;- create_genome(4, 100, 50)
ref$filter_chroms(90, "size")
ref$filter_chroms(0.4, "prop")

</pre><p></div></p>

<p><hr>
<a id="method-replace_Ns"></a></p><h3>Method <code>replace_Ns()</code></h3>
<p>Replace <code>N</code>s in the reference genome.</p><h3>Usage</h3>
<p><div class="r"></p><pre>ref_genome$replace_Ns(pi_tcag, n_threads = 1, show_progress = FALSE)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>pi_tcag</code></dt><dd><p>Numeric vector (length 4) indicating the sampling weights
for <code>T</code>, <code>C</code>, <code>A</code>, and <code>G</code>, respectively, for generating new nucleotides
with which to replace the <code>N</code>s.</p></dd>

<dt><code>n_threads</code></dt><dd><p>Optional integer specifying the threads to use.
Ignored if the package wasn't compiled with OpenMP. Defaults to <code>1</code>.</p></dd>

<dt><code>show_progress</code></dt><dd><p>Optional logical indicating whether to show a
progress bar. Defaults to <code>FALSE</code>.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>



    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$set_names`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>10</span>)
<span class='no'>ref</span>$<span class='fu'>set_names</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"a"</span>, <span class='st'>"b"</span>, <span class='st'>"c"</span>, <span class='st'>"d"</span>))


<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$clean_names`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>10</span>)
<span class='no'>ref</span>$<span class='fu'>set_names</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"a:"</span>, <span class='st'>"b|"</span>, <span class='st'>"c;"</span>, <span class='st'>"d'"</span>))
<span class='no'>ref</span>$<span class='fu'>clean_names</span>()


<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$add_chroms`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>10</span>)
<span class='no'>ref</span>$<span class='fu'>add_chroms</span>(<span class='st'>"TCAGTCAG"</span>)

<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$rm_chroms`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>10</span>)
<span class='no'>ref</span>$<span class='fu'>set_names</span>(<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"a"</span>, <span class='st'>"b"</span>, <span class='st'>"c"</span>, <span class='st'>"d"</span>))
<span class='no'>ref</span>$<span class='fu'>rm_chroms</span>(<span class='st'>"b"</span>)


<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$merge_chroms`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>10</span>)
<span class='no'>ref</span>$<span class='fu'>merge_chroms</span>()


<span class='co'>## ------------------------------------------------</span>
<span class='co'>## Method `ref_genome$filter_chroms`</span>
<span class='co'>## ------------------------------------------------</span>

<span class='no'>ref</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='create_genome.html'>create_genome</a></span>(<span class='fl'>4</span>, <span class='fl'>100</span>, <span class='fl'>50</span>)
<span class='no'>ref</span>$<span class='fu'>filter_chroms</span>(<span class='fl'>90</span>, <span class='st'>"size"</span>)
<span class='no'>ref</span>$<span class='fu'>filter_chroms</span>(<span class='fl'>0.4</span>, <span class='st'>"prop"</span>)</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#details">Details</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#methods">Methods</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


