<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>An R6 Class Representing Haploid Variants — variants • jackalope</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="An R6 Class Representing Haploid Variants — variants" />
<meta property="og:description" content="Interactive wrapper for a pointer to a C++ object that stores information about
haploid variants from a single reference genome." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/lucasnell/jackalope">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>An R6 Class Representing Haploid Variants</h1>
    <small class="dont-index">Source: <a href='https://github.com/lucasnell/jackalope/blob/master/R/aaa-classes.R'><code>R/aaa-classes.R</code></a></small>
    <div class="hidden name"><code>variants.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interactive wrapper for a pointer to a C++ object that stores information about
haploid variants from a single reference genome.</p>
    </div>



    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This class should NEVER be created using <code>variants$new</code>.
Only use <code>create_variants</code>.
Because this class wraps a pointer to a C++ object, there are no fields to
manipulate directly.
All manipulations are done through this class's methods.</p>
    <h2 class="hasAnchor" id="connections-to-ref-genome-objects"><a class="anchor" href="#connections-to-ref-genome-objects"></a>Connections to <code>ref_genome</code> objects</h2>

    

<p>Regarding the <code>ref_genome</code> object you use to create a <code>variants</code> object, you should
note the following:</p>
<ul>
<li><p><strong>This point is the most important.</strong>
Both the <code>ref_genome</code> and <code>variants</code> objects use the same underlying
C++ object to store reference genome information.
Thus, if you make any changes to the <code>ref_genome</code> object, those changes will
also show up in the <code>variants</code> object.
For example, if you make a <code>variants</code> object named <code>V</code>
based on an existing <code>ref_genome</code> object named <code>R</code>,
then you merge chromosomes in <code>R</code>,
<code>V</code> will now have merged chromosomes.
If you've already started adding mutations to <code>V</code>,
then all the indexes used to store those mutations will be inaccurate.
So when you do anything with <code>V</code> later, your R session will crash
or have errors.
<strong>The lesson here is that you shouldn't edit the reference
genome after using it to create variants.</strong></p></li>
<li><p>If a <code>ref_genome</code> object is used to create a <code>variants</code>
object, deleting the <code>ref_genome</code> object won't cause issues with
the <code>variants</code> object.
However, the <code>variants</code> class doesn't provide methods to edit
chromosomes, so only remove the <code>ref_genome</code> object when you're done
editing the reference genome.</p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='create_variants.html'>create_variants</a></code></p></div>
    <h2 class="hasAnchor" id="methods"><a class="anchor" href="#methods"></a>Methods</h2>

    
<h3>Public methods</h3>

<ul>
<li><p><a href='#method-new'><code>variants$new()</code></a></p></li>
<li><p><a href='#method-print'><code>variants$print()</code></a></p></li>
<li><p><a href='#method-ptr'><code>variants$ptr()</code></a></p></li>
<li><p><a href='#method-n_chroms'><code>variants$n_chroms()</code></a></p></li>
<li><p><a href='#method-n_vars'><code>variants$n_vars()</code></a></p></li>
<li><p><a href='#method-sizes'><code>variants$sizes()</code></a></p></li>
<li><p><a href='#method-chrom_names'><code>variants$chrom_names()</code></a></p></li>
<li><p><a href='#method-var_names'><code>variants$var_names()</code></a></p></li>
<li><p><a href='#method-chrom'><code>variants$chrom()</code></a></p></li>
<li><p><a href='#method-gc_prop'><code>variants$gc_prop()</code></a></p></li>
<li><p><a href='#method-nt_prop'><code>variants$nt_prop()</code></a></p></li>
<li><p><a href='#method-set_names'><code>variants$set_names()</code></a></p></li>
<li><p><a href='#method-add_vars'><code>variants$add_vars()</code></a></p></li>
<li><p><a href='#method-dup_vars'><code>variants$dup_vars()</code></a></p></li>
<li><p><a href='#method-rm_vars'><code>variants$rm_vars()</code></a></p></li>
<li><p><a href='#method-add_sub'><code>variants$add_sub()</code></a></p></li>
<li><p><a href='#method-add_ins'><code>variants$add_ins()</code></a></p></li>
<li><p><a href='#method-add_del'><code>variants$add_del()</code></a></p></li>
</ul>
<p><hr>
<a id="method-new"></a></p><h3>Method <code>new()</code></h3>
<p>Do NOT use this; only use <code>create_variants</code> to make new <code>variants</code>.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$new(genomes_ptr, reference_ptr)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>genomes_ptr</code></dt><dd><p>An <code>externalptr</code> object pointing to a C++ object that
stores the information about the variants.</p></dd>

<dt><code>reference_ptr</code></dt><dd><p>An <code>externalptr</code> object pointing to a C++ object that
stores the information about the reference genome.</p></dd>

</dl><p></div></p>
<p><hr>
<a id="method-print"></a></p><h3>Method <code><a href='https://rdrr.io/r/base/print.html'>print()</a></code></h3>
<p>Print a <code>variants</code> object.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$print()</pre><p></div></p>

<p><hr>
<a id="method-ptr"></a></p><h3>Method <code>ptr()</code></h3>
<p>View pointer to underlying C++ object (this is not useful to end users).</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$ptr()</pre><p></div></p>

<h3>Returns</h3>
<p>An <code>externalptr</code> object.</p>
<p><hr>
<a id="method-n_chroms"></a></p><h3>Method <code>n_chroms()</code></h3>
<p>View number of chromosomes.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$n_chroms()</pre><p></div></p>

<h3>Returns</h3>
<p>Integer number of chromosomes.</p>
<p><hr>
<a id="method-n_vars"></a></p><h3>Method <code>n_vars()</code></h3>
<p>View number of variants.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$n_vars()</pre><p></div></p>

<h3>Returns</h3>
<p>Integer number of variants.</p>
<p><hr>
<a id="method-sizes"></a></p><h3>Method <code>sizes()</code></h3>
<p>View chromosome sizes for one variant.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$sizes(var_ind)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>Integer vector of chromosome sizes for focal variant.</p>
<p><hr>
<a id="method-chrom_names"></a></p><h3>Method <code>chrom_names()</code></h3>
<p>View chromosome names.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$chrom_names()</pre><p></div></p>

<h3>Returns</h3>
<p>Character vector of chromosome names.</p>
<p><hr>
<a id="method-var_names"></a></p><h3>Method <code>var_names()</code></h3>
<p>View variant names.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$var_names()</pre><p></div></p>

<h3>Returns</h3>
<p>Character vector of variant names.</p>
<p><hr>
<a id="method-chrom"></a></p><h3>Method <code>chrom()</code></h3>
<p>View one variant chromosome.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$chrom(var_ind, chrom_ind)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A single string representing the chosen variant chromosome's DNA
sequence.</p>
<p><hr>
<a id="method-gc_prop"></a></p><h3>Method <code>gc_prop()</code></h3>
<p>View GC proportion for part of one variant chromosome.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$gc_prop(var_ind, chrom_ind, start, end)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>start</code></dt><dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>

<dt><code>end</code></dt><dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is either <code>G</code> or <code>C</code>.</p>
<p><hr>
<a id="method-nt_prop"></a></p><h3>Method <code>nt_prop()</code></h3>
<p>View nucleotide content for part of one variant chromosome</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$nt_prop(nt, var_ind, chrom_ind, start, end)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>nt</code></dt><dd><p>Which nucleotide to calculate the proportion that the DNA
sequence is made of. Must be one of <code>T</code>, <code>C</code>, <code>A</code>, <code>G</code>, or <code>N</code>.</p></dd>

<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>start</code></dt><dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>

<dt><code>end</code></dt><dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is <code>nt</code>.</p>
<p><hr>
<a id="method-set_names"></a></p><h3>Method <code>set_names()</code></h3>
<p>Change variant names.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$set_names(new_names)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>new_names</code></dt><dd><p>Vector of new names to use. This must be the same length as
the number of current names.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<p><hr>
<a id="method-add_vars"></a></p><h3>Method <code>add_vars()</code></h3>
<p>Add one or more blank, named variants</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$add_vars(new_names)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>new_names</code></dt><dd><p>Vector of name(s) for the new variant(s).</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<p><hr>
<a id="method-dup_vars"></a></p><h3>Method <code>dup_vars()</code></h3>
<p>Duplicate one or more variants by name.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$dup_vars(var_names, new_names = NULL)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_names</code></dt><dd><p>Vector of existing variant name(s) that you want to
duplicate.</p></dd>

<dt><code>new_names</code></dt><dd><p>Optional vector specifying the names of the duplicates.
If <code>NULL</code>, their names are auto-generated. Defaults to <code>NULL</code>.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<p><hr>
<a id="method-rm_vars"></a></p><h3>Method <code>rm_vars()</code></h3>
<p>Remove one or more variants by name.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$rm_vars(var_names)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_names</code></dt><dd><p>Vector of existing variant name(s) that you want to remove.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<p><hr>
<a id="method-add_sub"></a></p><h3>Method <code>add_sub()</code></h3>
<p>Manually add a substitution.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$add_sub(var_ind, chrom_ind, pos, nt)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>pos</code></dt><dd><p>Position at which to add the mutation.</p></dd>

<dt><code>nt</code></dt><dd><p>Single character representing the nucleotide to change the
current one to.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>
<p><hr>
<a id="method-add_ins"></a></p><h3>Method <code>add_ins()</code></h3>
<p>Manually add an insertion.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$add_ins(var_ind, chrom_ind, pos, nts)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>pos</code></dt><dd><p>Position at which to add the mutation.</p></dd>

<dt><code>nts</code></dt><dd><p>String representing the nucleotide(s) that will be inserted after
the designated position.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p><pre>\item{`add_del(var_ind, chrom_ind, pos, n_nts)`}{Manually add a deletion
    for a given variant (`var_ind`), chromosome (`chrom_ind`), and position (`pos`).
    The designated number of nucleotides to delete (`n_nts`) will be deleted
    starting at `pos`, unless `pos` is near the chromosome end and doesn't have
    `n_nts` nucleotides to remove; it simply stops at the chromosome end in
    this case.}
</pre>

<p><hr>
<a id="method-add_del"></a></p><h3>Method <code>add_del()</code></h3>
<p>Manually add a deletion.</p><h3>Usage</h3>
<p><div class="r"></p><pre>variants$add_del(var_ind, chrom_ind, pos, n_nts)</pre><p></div></p>

<h3>Arguments</h3>
<p><div class="arguments"></p><dl class='dl-horizontal'>
<dt><code>var_ind</code></dt><dd><p>Index for the focal variant.</p></dd>

<dt><code>chrom_ind</code></dt><dd><p>Index for the focal chromosome.</p></dd>

<dt><code>pos</code></dt><dd><p>Position at which to add the mutation.</p></dd>

<dt><code>n_nts</code></dt><dd><p>Single integer specifying the number of nucleotides to delete.
These will be deleted starting at <code>pos</code>.
If <code>pos</code> is near the chromosome end and doesn't have <code>n_nts</code> nucleotides
to remove, it simply removes nucleotides from <code>pos</code> to the chromosome end.</p></dd>

</dl><p></div></p>
<h3>Returns</h3>
<p>This <code>R6</code> object, invisibly.</p>



  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#details">Details</a></li>
      <li><a href="#connections-to-ref-genome-objects">Connections to <code>ref_genome</code> objects</a></li>
      <li><a href="#see-also">See also</a></li>
      <li><a href="#methods">Methods</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


